homeassistant/components/device_tracker/setup.py

```diff
@@ -0,0 +1,199 @@
+        """Set up a legacy platform."""
+        LOGGER.info("Setting up %s.%s", DOMAIN, self.type)
+        try:
+            scanner = None
+            setup = None
+            if hasattr(self.platform, 'async_get_scanner'):
+                scanner = await self.platform.async_get_scanner(
+                    hass, {DOMAIN: self.config})
+            elif hasattr(self.platform, 'get_scanner'):
+                scanner = await hass.async_add_job(
+                    self.platform.get_scanner, hass, {DOMAIN: self.config})
+            elif hasattr(self.platform, 'async_setup_scanner'):
+                setup = await self.platform.async_setup_scanner(
+                    hass, self.config, tracker.async_see, discovery_info)
+            elif hasattr(self.platform, 'setup_scanner'):
+                setup = await hass.async_add_job(
+                    self.platform.setup_scanner, hass, self.config,
+                    tracker.see, discovery_info)
+            elif hasattr(self.platform, 'async_setup_entry'):
+                setup = await self.platform.async_setup_entry(
+                    hass, self.config, tracker.async_see)
+            else:
+                raise HomeAssistantError(
+                    "Invalid legacy device_tracker platform.")
+
+            if scanner:
+                async_setup_scanner_platform(
+                    hass, self.config, scanner, tracker.async_see, self.type)
+                return
+
+            if not setup:
+                LOGGER.error("Error setting up platform %s", self.type)
+                return
+
+        except Exception:  # pylint: disable=broad-except
+            LOGGER.exception("Error setting up platform %s", self.type)
+
+
+async def async_extract_config(hass, config):
+    """Extract device tracker config and split between legacy and modern."""
+    legacy = []
+    entity_platform = []
+
+    for platform in await asyncio.gather(*[
+            async_create_platform_type(hass, config, p_type, p_config)
+            for p_type, p_config in config_per_platform(config, DOMAIN)
+    ]):
+        if platform is None:
+            continue
+
+        if platform.type == PLATFORM_TYPE_ENTITY:
+            entity_platform.append(platform)
+        elif platform.type == PLATFORM_TYPE_LEGACY:
+            legacy.append(platform)
+        else:
+            raise ValueError("Unable to determine type for {}: {}".format(
+                platform.name, platform.type))
+
+    return (legacy, entity_platform)
+
+
+async def async_create_platform_type(hass, config, p_type, p_config) \
+        -> Optional[DeviceTrackerPlatform]:
+    """Determine type of platform."""
+    platform = await async_prepare_setup_platform(
+        hass, config, DOMAIN, p_type)
+
+    if platform is None:
+        return None
+
+    return DeviceTrackerPlatform(p_type, platform, p_config)
+
+
+@callback
+def async_setup_scanner_platform(hass: HomeAssistantType, config: ConfigType,
```

Comment:
This is only for the legacy type, right? Should we move it to legacy?

Reply:
Future PR 👍